### PR TITLE
Refactor AuthorityType functionality

### PR DESCRIPTION
**Pull request by SameerK-MSFT**
Fixes # 

**Changes proposed in this request**
Shifting the functionality to Authority and derived classes would be the ideal way to refactor and have O in the code. But due to the tight coupling and functionality spread over various segments, that will require significant efforts. In the interest of time, it has a few changes.

Here is the summary of changes:

- Many callers are using AuthorityInfo.AuthorityType to determine the behavior. The behavior + functionality of a type of Authority has been encapsulated in the AuthorityInfo in properties.
- There is a new class AuthorityInfoHelper. This class extends the functionality of AuthorityInfo. It is intentionally a nested class inside AuhtorityInfo for the following reasons:
  - The callers have access to AuthorityInfo; but not Authority itself. So the functionality needs to be wrapped inside AuthorityInfo itself.
  - AuthoirityInfo should not be bloated. This class provides the extended functionality.
  - It has to be nested to access the private members of the AuthorityInfo
  - Ideally this should be nested class of Authority itself. But since it requires access to the private members of Authority and the callers lack access to Authority, it has to be nested class of AuthorityInfo
- To centralize the extended functionality, some helper static functionality has been shifted to AuthorityInfo. In order to minimize the impact on the callers, it Authority delegates that to AuthorityInfo
 - AuthorityValidatorFactory was a helper class with single method to create. Functionality has been shifted to AuthorityInoHelper.
- Changed the AuthorityType from public to internal. Making it private would require many more changes even in test. So this may be the intermediate solution to increase the O factor.

**Testing**
Existing tests

**Performance impact**
None

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 1]
```diff
@@ -87,10 +116,18 @@ internal class AuthorityInfo
 
         public string Host { get; }
         public string CanonicalAuthority { get; }
-        public AuthorityType AuthorityType { get; }
+        internal AuthorityType AuthorityType { get; }
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:119]: Shouldn't matter since enclosing AuthorityInfo is internal?

**Review by pmaytak — COMMENTED** [review]
nit: Can you set the dsts branch as base for this review, so the dsts changes are excluded?

Reviewed partially, will finish tomorrow.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/DstsTests.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+                result = await app
+                    .AcquireTokenForClient(TestConstants.s_scope)
+                    .ExecuteAsync(CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                Assert.IsNotNull(result);
+                Assert.IsNotNull(result.AccessToken);
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);    
+                //request to token endpoint
```
Comment (neha-bhargava) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/DstsTests.cs:93]: This comment can be removed. I added this as reminder to assert it and added the checks above at line 47 but forgot to remove this.

**Review by neha-bhargava — DISMISSED** [review]

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 3]
```diff
@@ -87,10 +116,18 @@ internal class AuthorityInfo
...
         public string UserRealmUriPrefix { get; }
         public bool ValidateAuthority { get; }
 
+        internal bool IsInstanceDiscoverySupported => (AuthorityType == AuthorityType.Aad);
+
+        internal bool IsClientInfoSupported => (AuthorityType == AuthorityType.Aad || AuthorityType == AuthorityType.Dsts);
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:125]: Should include B2C.

**Review by pmaytak — APPROVED** [review]
Didn't review DSTS code.

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs (hunk 3 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs]: Now includes B2C in the latest check-in

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs (hunk 1 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:119]: Just getting closer to O.

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 4]
```diff
@@ -112,12 +116,15 @@ private string[] GetPathSegments(string absolutePath)
 
         public string Host { get; }
         public string CanonicalAuthority { get; }
-        public AuthorityType AuthorityType { get; }
+        internal AuthorityType AuthorityType { get; }
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:119]: Maybe add a comment saying that this property should be removed.

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**